src/mol-math/geometry/symmetry-operator.ts

```diff
@@ -167,127 +168,123 @@ namespace SymmetryOperator {
-        return generalPosition(t, coords);
-    }
-}
+        x(i: T): number {
+            const m = this._m;
+            const xx = m[0], yy = m[4], zz = m[8], tx = m[12];
 
-export { SymmetryOperator };
+            const x = this._x[i], y = this._y[i], z = this._z[i], w = (m[3] * x + m[7] * y + m[11] * z + m[15]) || 1.0;
+            return (xx * x + yy * y + zz * z + tx) / w;
+        }
 
-function _zeroRadius(i: number) { return 0; }
+        y(i: T): number {
+            const m = this._m;
+            const xx = m[1], yy = m[5], zz = m[9], ty = m[13];
 
-interface Projections { x(index: number): number, y(index: number): number, z(index: number): number }
+            const x = this._x[i], y = this._y[i], z = this._z[i], w = (m[3] * x + m[7] * y + m[11] * z + m[15]) || 1.0;
+            return (xx * x + yy * y + zz * z + ty) / w;
+        }
 
-function createProjections(t: SymmetryOperator, coords: SymmetryOperator.Coordinates): Projections {
-    if (t.isIdentity) return { x: projectCoord(coords.x), y: projectCoord(coords.y), z: projectCoord(coords.z) };
-    return { x: projectX(t, coords), y: projectY(t, coords), z: projectZ(t, coords) };
-}
+        z(i: T): number {
+            const m = this._m;
+            const xx = m[2], yy = m[6], zz = m[10], tz = m[14];
 
-function projectCoord(xs: ArrayLike<number>) {
-    return function projectCoord(i: number) {
-        return xs[i];
-    };
-}
+            const x = this._x[i], y = this._y[i], z = this._z[i], w = (m[3] * x + m[7] * y + m[11] * z + m[15]) || 1.0;
+            return (xx * x + yy * y + zz * z + tz) / w;
+        }
+    }
 
-function isW1(m: Mat4) {
-    return m[3] === 0 && m[7] === 0 && m[11] === 0 && m[15] === 1;
-}
+    class _ArrayMappingW1<T extends number> implements ArrayMapping<T> {
+        private readonly _x: ArrayLike<number>;
+        private readonly _y: ArrayLike<number>;
+        private readonly _z: ArrayLike<number>;
+        private readonly _m: Mat4;
+
+        constructor(readonly operator: SymmetryOperator, readonly coordinates: Coordinates, readonly r: ((index: T) => number) = _zeroRadius) {
+            this._x = coordinates.x;
+            this._y = coordinates.y;
+            this._z = coordinates.z;
+            this._m = operator.matrix;
+        }
 
-function projectX({ matrix: m }: SymmetryOperator, { x: xs, y: ys, z: zs }: SymmetryOperator.Coordinates) {
-    const xx = m[0], yy = m[4], zz = m[8], tx = m[12];
+        invariantPosition(i: T, s: Vec3): Vec3 {
+            s[0] = this._x[i];
+            s[1] = this._y[i];
+            s[2] = this._z[i];
+            return s;
+        }
 
-    if (isW1(m)) {
-        // this should always be the case.
-        return function projectX_W1(i: number) {
-            return xx * xs[i] + yy * ys[i] + zz * zs[i] + tx;
-        };
-    }
+        position(i: T, s: Vec3): Vec3 {
+            const m = this._m;
+            const x = this._x[i], y = this._y[i], z = this._z[i];
```

Comment:
In this particular case, I think actually doing a local copy isn't beneficial compared to using `this.` everywhere. The only difference would if it was caching nested prop (`this.x.y`).

It should be fine to drop the `_` and just use `this.x` etc everywhere.

Reply:
nice, that helps quite some